tests/flow/test_hnswlib.py

```diff
@@ -518,11 +450,522 @@ def test_multi_range_query():
+              f" {total_correct_cur_chunk/(k*chunk_size)}")
+
+
+def test_parallel_batch_search():
+    dim = 32
+    num_elements = 100000
+    num_queries = 5000
+    batch_size = 100
+    n_batches = 5
+    n_threads = min(os.cpu_count(), 8)
+    expected_parallel_rate = 0.85  # we expect that at least 85% of the insert/search time will be executed in parallel
+    expected_speedup = 1 / ((1-expected_parallel_rate) + expected_parallel_rate/n_threads)  # by Amdahl's law
+
+    index = create_hnsw_index(dim, num_elements, VecSimMetric_Cosine, VecSimType_FLOAT32, ef_runtime=200)
+
+    bf_params = BFParams()
+
+    bf_params.initialCapacity = num_elements
+    bf_params.blockSize = num_elements
+    bf_params.dim = dim
+    bf_params.type = VecSimType_FLOAT32
+    bf_params.metric = VecSimMetric_Cosine
+
+    bf_index = BFIndex(bf_params)
+
+    # Load BF and HNSW indexes with data
+    np.random.seed(47)
+    data = np.float32(np.random.random((num_elements, dim)))
+    index.add_vector_parallel(data, np.array(range(num_elements)), n_threads)
+    for i, vector in enumerate(data):
+        bf_index.add_vector(vector, i)
+
+    query_data = np.float32(np.random.random((num_queries, dim)))
+
+    # Sequential batched search as the baseline
+    total_search_time = 0
+    total_correct = 0
+    total_res_bf = []  # save the ground truth
+    for i, query in enumerate(query_data):
+        start = time.time()
+        batch_iterator = index.create_batch_iterator(query)
+        # Collect all the results from all batches
+        res_labels = set()
+        for _ in range(n_batches):
+            res_labels = res_labels.union(set(batch_iterator.get_next_results(batch_size, BY_SCORE)[0][0]))
+
+        total_search_time += time.time() - start
+        res_labels_bf, _ = bf_index.knn_query(query, n_batches * batch_size)
+        total_res_bf.append(set(res_labels_bf[0]))
+        total_correct += len(res_labels.intersection(set(res_labels_bf[0])))
+
+    print(f"Running sequential batched search of {n_batches} batches of size {batch_size}, over {num_queries} queries,"
+          f" got recall of {total_correct/(n_batches*batch_size*num_queries)} and average query time is"
+          f" {total_search_time / num_queries} seconds")
+
+    total_results_parallel = {}
+
+    def run_batched_search(query_, query_ind):
+        batch_iterator_ = index.create_batch_iterator(query_)
+        res_labels_ = set()
+        for _ in range(n_batches):
+            res_labels_ = res_labels_.union(set(batch_iterator_.get_next_results(batch_size, BY_SCORE)[0][0]))
+        total_results_parallel[query_ind] = res_labels_
+
+    start = time.time()
+    with ThreadPoolExecutor(max_workers=n_threads) as executor:
+        futures = [executor.submit(run_batched_search, q, i) for i, q in enumerate(query_data)]
+        done, not_done = wait(futures, return_when=concurrent.futures.ALL_COMPLETED)
```

Comment:
I think you can create a function that gets a collection of batch iterators and runs them in parallel, same as with KNN and range.

Reply:
I know, but I preferred doing it that way for the batch iterator. I'll explain when we'll talk